examples/step-19/step-19.cc

```diff
@@ -0,0 +1,728 @@
+                                       update_gradients |
+                                       update_normal_vectors);
+    std::vector<Tensor<1, dim>> solution_gradients(
+      fe_face_values.n_quadrature_points);
+    const FEValuesExtractors::Scalar electric_potential(0);
+
+    types::particle_index n_current_particles =
+      particle_handler.n_global_particles();
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      for (const auto &face : cell->face_iterators())
+        if (face->at_boundary() && face->boundary_id() == BoundaryIds::cathode)
+          {
+            // compute the electric field strength at each quadrature point
+            fe_face_values.reinit(cell, face);
+            fe_face_values[electric_potential].get_function_gradients(
+              solution, solution_gradients);
+            for (const unsigned int q_point :
+                 fe_face_values.quadrature_point_indices())
+              {
+                const Tensor<1, dim> E = solution_gradients[q_point];
+
+                // electrons can only escape the cathode if the electric field
+                // strength exceeds a threshold and,
+                // crucially, if the electric field points *into* the domain
+                if ((E * fe_face_values.normal_vector(q_point) < 0) &&
+                    (E.norm() > Constants::E_threshold))
+                  {
+                    const Point<dim> location =
+                      fe_face_values.quadrature_point(q_point);
+                    // create a particle at 'location' and insert it into the
+                    // ParticleHandler
+
+                    Particles::Particle<dim> new_particle;
+                    new_particle.set_location(location);
+                    new_particle.set_reference_location(
+                      mapping.transform_real_to_unit_cell(cell, location));
+                    new_particle.set_id(n_current_particles);
+                    particle_handler.insert_particle(new_particle, cell);
+
+                    ++n_current_particles;
+                  }
+              }
+          }
+
+    particle_handler.update_cached_numbers();
+  }
+
+
+  // @sect4{CathodeRaySimulator::move_particles}
+
+  template <int dim>
+  void CathodeRaySimulator<dim>::move_particles()
+  {
+    // advance all particles
+    const double dt = time.get_next_step_size();
+
+    std::vector<Point<dim>>     particle_positions;
+    std::vector<Tensor<1, dim>> field_gradients;
+
+    // Loop over all cells and advect the particles cell-wise
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      if (particle_handler.n_particles_in_cell(cell) > 0)
+        {
+          const typename Particles::ParticleHandler<
+            dim>::particle_iterator_range particles_in_cell =
+            particle_handler.particles_in_cell(cell);
+
+          const unsigned int n_particles_in_cell =
+            particle_handler.n_particles_in_cell(cell);
+
```

Comment:
I do not think we should make too many optimizations for a simple step, but since this code will likely be copied into many programs as the canonical way to do things, can we move this call outside of the `if(n_particles_in_cell > 0)`? It is of `O(NlogN)` complexity which may matter for millions of particles. In fact the optimal way of writing these lines would probably be:
```suggestion
    for (const auto &cell : dof_handler.active_cell_iterators())
    {
     const typename Particles::ParticleHandler<
            dim>::particle_iterator_range particles_in_cell =
            particle_handler.particles_in_cell(cell);

          const unsigned int n_particles_in_cell =
            std::distance(particles_in_cell.begin(),particles_in_cell.end());
      if (n_particles_in_cell > 0)
        {
         

```

Reply:
I'm happy to do this, but shouldn't the underlying call to `equal_range` just be `O(2*log(N))`? That's one `log N` for the lower-bound for the key, and another one for the upper-bound?

Or do you mean that we do this `M` times (for `M` = number of cells) that that's how you get to your complexity?

Reply:
I meant that at the moment there are 3 calls that are executed `n_cells` times with `O(log(N))` each (2 calls to `n_particles_in_cell` and one to `particles_in_cell`), while in practice only one call is necessary, namely the one to `particles_in_cell`. `n_particles_in_cell` can then be computed with `std::distance` (linear in distance). Changing to my suggested code would not change the overall complexity (`O(NlogN)`), but would reduce the number of calls with that complexity.

Reply:
Good point. I've taken this as a teaching opportunity and instead just doing as you say, pointed out that it's inefficient and wrote a section in the "Possibilities for extensions" about it, and in particular how to reduce the effort to O(N).